### PR TITLE
Add use-last parameter

### LINKED ISSUE #13: Allow alternate active selection
Some users have expressed the desire to select the last heading on the current page over what hydra currently considers the active element, if one exists.

### PR DESCRIPTION
**Pull request by freundTech**
Closes #13.
This allows hydra to alternatively display the last instead of the first candidate on a page, mimicking the behavior of LaTeX.

### CODE REVIEW

File: src/core.typ [hunk 1]
```diff
@@ -85,38 +85,42 @@
   let next-ancestor = none
 
   if ctx.ancestors != none {
-    let prev = query(selector(ctx.ancestors.target).before(ctx.anchor-loc))
-    let next = query(selector(ctx.ancestors.target).after(ctx.anchor-loc))
+    let prev-ancestors = query(selector(ctx.ancestors.target).before(ctx.anchor-loc))
+    let next-ancestors = query(selector(ctx.ancestors.target).after(ctx.anchor-loc))
```
Comment (freundTech) [inline src/core.typ:89]: I changed some of the variable names to improve readability and code-flow. Previously the same variable was used for multiple things, which would have required me to change code flow (because I needed access to variables that were already overwritten)

File: src/core.typ [hunk 2]
```diff
@@ -85,38 +85,42 @@
...
+  let prev-targets = query(look-behind)
+  let next-targets = query(look-ahead)
 
   if ctx.primary.filter != none {
-    prev = prev.filter(x => (ctx.primary.filter)(ctx, x))
-    next = next.filter(x => (ctx.primary.filter)(ctx, x))
+    prev-targets = prev-targets.filter(x => (ctx.primary.filter)(ctx, x))
+    next-targets = next-targets.filter(x => (ctx.primary.filter)(ctx, x))
   }
+  next-targets = next-targets.filter(x => x.location().page() == ctx.anchor-loc.page())
```
Comment (freundTech) [inline src/core.typ:116]: This could be put behind a `if ctx.use-last` guard, but I don't think it's necessary. I can't think of any case where you would want hydra to display a heading that is defined on the *next* page.

File: src/core.typ [hunk 3]
```diff
@@ -85,38 +85,42 @@
...
 
-  let prev = if prev != () { prev.last() }
-  let next = if next != () { next.first() }
+  let prev = if prev-targets != () { prev-targets.last() }
+  let next = if next-targets != () { next-targets.first() }
+  let last = if next-targets != () { next-targets.last() }
 
   (
-    primary: (prev: prev, next: next),
+    primary: (prev: prev, next: next, last: last),
```
Comment (freundTech) [inline src/core.typ:123]: I thought about just setting `next: last` if `ctx.use-last` is set, but that breaks the redundancy checking. That still requires `next`, even if `use-last` is set.

File: src/core.typ [hunk 4]
```diff
@@ -227,11 +231,20 @@
...
+      (ctx.display)(ctx, candidates.primary.last)
+    } else if prev-eligible {
+      (ctx.display)(ctx, candidates.primary.prev)
+    }
+  } else if not ctx.use-last {
+    if prev-eligible and not active-redundant {
+      (ctx.display)(ctx, candidates.primary.prev)
+    } else if next-eligible and not ctx.skip-starting {
+      (ctx.display)(ctx, candidates.primary.next)
+    }
```
Comment (freundTech) [inline src/core.typ:248]: I would have liked to simplify this a bit, but I couldn't figure out a simpler solution.

**Comment by freundTech** [comment]
Noticed that my PR was already outdated when I created it. Just rebased and force-pushed.

File: src/core.typ (hunk 3 above)
Comment (tingerrr) [reply src/core.typ:127]: Yeah, that makes sense. This must still be documented in the manual in the custom types section.

File: src/core.typ (hunk 4 above)
Comment (tingerrr) [reply src/core.typ:252]: This one's tricky, it was already a little hard to follow before. I think you can invert a few conditions and use early `return` to make this easier to read.

File: src/lib.typ [hunk 5]
```diff
@@ -28,6 +28,8 @@
 ///   display. If this is `auto`, the default implementaion will be used.
 /// - skip-starting (bool): Whether `hydra` should show the current candidate even if it's on top of
 ///   the current page.
+/// - use-last (bool): If hydra should show the name of the first of last candidate on the page.
+//    Defaults to false.
```
Comment (tingerrr) [inline src/lib.typ:32]: A little slip up here, I assume.

**Review by tingerrr — CHANGES_REQUESTED** [review]
A few tiny changes and it's good to go! Surprisingly simple solution as it turns out.

File: src/core.typ (hunk 3 above)
Comment (freundTech) [reply src/core.typ:128]: Done

**Comment by freundTech** [comment]
I reworked it a bit.
`get-candidates` no longer considers `ctx.use-last`, but always returns correct `prev`, `next` and `last`, while `execute` chooses which of these to display.

`execute` is now significantly easier.

**Review by tingerrr — APPROVED** [review]
LGTM, Thanks!

**Comment by freundTech** [comment]
Are you planning to make a release for this or are you waiting for more changes?
I would like to stop vendoring hydra into my thesis :wink:

**Comment by tingerrr** [comment]
I'll prepare some releases today, since I have to fix some stuff on subpar. A shame I couldn't get the mantys PR ready in time, but it is what it is.
